### PR TITLE
Link Filters/Sorting + UTM Builder + Password Protected Links + Auto Expiring Links

### PR DESCRIPTION
**Pull request by steven-tey**
A few issues this PR fixes:
- Password-protected links: https://github.com/steven-tey/dub/issues/34
- Auto-expiring links: https://github.com/steven-tey/dub/issues/48
- Add a UTM builder / opengraph features: https://www.trustpilot.com/reviews/63487fd9377e522e314e76db

TODO:
- [x] Refactor to use MySQL as the primary database for storing link metadata (instead of Redis)
- [x] Password protection
- [x] UTM builder
- [x] Auto Expiring links
- [x] Custom OG tags
- [x] Archive links
- [x] Link Filters
- [x] Link Sorting

### CODE REVIEW

**Comment by steven-tey** [comment]
Added password protection feature for links 🔒 🥳 (fixes #34)

https://user-images.githubusercontent.com/28986134/196017168-df62d6e6-522d-48eb-803f-001c357c96ac.mp4

Also:
- Standardized UI for login, register, and password protection pages
- Installed bcrypt
- Updated README for new links structure
- Updated prisma schema for links

File: components/app/modals/archive-link-modal.tsx [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+                  : `/api/links/${props.key}/archive`,
+                {
+                  method: "PUT",
+                  headers: {
+                    "Content-Type": "application/json",
+                  },
+                },
+              ).then(async (res) => {
+                setArchiving(false);
+                console.log(res.status, "mutating");
```
Comment (QuiiBz) [inline components/app/modals/archive-link-modal.tsx]: You might have forgotten to remove this `console.log` 😅

Comment (steven-tey) [reply components/app/modals/archive-link-modal.tsx]: Oh wow, great catch, removing it now! Thank you @QuiiBz!! 🙏

**Comment by steven-tey** [comment]
MERGING!!!!! 🤩😱

**Pull request merged**